experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+//!     pub enum Style {
+//!         Long, Short, Narrow,
+//!     }
+//!     pub enum Type {
+//!         Conjunction, Disjunction,
+//!     }
+//!     /// These are the options that go in.
+//!     /// Use as `options::In`.  I happen to like short names on types that
+//!     /// have the same name as the mod they are in, to reduce repetition.
+//!     pub struct In {
+//!         pub style: Style,
+//!         /// `type` is a reserved word.  Better suggestions welcome.
+//!         pub in_type: Type,
+//!     }
+//! }
+//! ```
+//!
+//! You would refer to them by `listformat::options::Style::Long` and the like, which gives us
+//! non-repetitive and reusable names.  `use` clauses can be used to shorten if needed:
+//!
+//! ```rust ignore
+//! use listformat::options::Style;
+//! // ...
+//! ```
+//!
+//! ## Locale
+//!
+//! Passing in a locale specifier is the first nontrivial design decision.  This is since locales
+//! may be implemented in a number of different ways:
+//!
+//! 1. Is the locale data owned or borrowed?
+//! 1. Are locales fallible or infallible? I.e. are they *well-formed* or always *valid*.
+//! 1. Do locales come from a system repository or are they somehow provided by the end user? 
+//!
+//! In Rust, each of these concerns seems to ask for different interfaces, making general
+//! interfaces hard to formulate.  I welcome a correction here if I'm wrong.  The objective is to
+//! unify as many as possible of the above use cases in a trait that can express all of them.
+//!
+//! ### "Stringly typed" API
+//!
+//! One possibility is to require a string-like type (`AsRef<str>`).
+//!
+//! Pros:
+//!
+//! * This is the closest to the JavaScript API surface
+//! * It is the easiest one to implement.
+//!
+//! Cons:
+//!
+//! * It has the loosest guarantees as to the functionality it provides.
+//!
+//! Here's an example of how a stringly typed trait can be used.  We can specify a general
+//! constraint on an input parameter and get a functional API. In the example above it is
+//! `AsRef<str>`, but we could have it be `AsRef<T>` for some other suitably chosen trait `T`.
+//!
+//! ```
+//! /// This is the general trait.
+//! use std::convert::AsRef;
+//! pub trait Trait {
+//!     fn example(l: impl AsRef<str>);
+//! }
+//!
+//! /// Sample trait implementation.  We see below that two implementations
+//! /// can coexist.
+//! struct I{}
+//! impl Trait for I {
+//!     fn example(l: impl AsRef<str>) {
+//!         let s: &str = l.as_ref();
+//!         println!("result = {}", s);
+//!     }
+//! }
+//!
+//! /// One implementation (`&'static str`) and example use.
+//! let s: &'static str = "hello!";
+//! I::example("hello!");
+//!
+//! /// Another implementation (`struct J`) and example use.
+//! struct J {
+//!   rep: String,
+//! }
+//! impl AsRef<str> for J {
+//!     fn as_ref(&self) -> &str {
+//!         &self.rep
+//!     }
+//! }
+//! let j = J{ rep: "world!".to_string() };
+//! I::example(&j);
+//! ```
+//!
```

Comment:
I'm torn on that because of fallibility. I know you have a section about it later.

I generally believe that Rust caters to "localize failure" approach - if you have a function that takes a number of inputs that then it parses and each can fail, instead expose each bit, let it have a fallible constructor, and then pass around already resolved and valid bits.

From that perspective, a constructor from a string should be fallible (hence, `try_new` or `try_new_from_str`) rather than infallible if it takes a string.
On the other hand, I don't know how well our FFI will handle separate types, so we may want to have a constructor that takes strings.
On yet another, if we can't pass types around, then we're likely going to end up with a very awkward API because we won't be able to pass Options or PluralCategory to NumberFormat and so on.

So my *hope* would be that we can have a constructor that takes a `Locale`. Now, even that constructor imho should be fallible because we may not have data for that locale.

Reply:
> I generally believe that Rust caters to "localize failure" approach

SGTM

> From that perspective, a constructor from a string should be fallible (hence, `try_new` or `try_new_from_str`) rather than infallible if it takes a string.
> On the other hand, I don't know how well our FFI will handle separate types, so we may want to have a constructor that takes strings.

> On yet another, if we can't pass types around, then we're likely going to end up with a very awkward API because we won't be able to pass Options or PluralCategory to NumberFormat and so on.

I suspect there is no good way to make an FFI-aware API that "feels elegant" in regular use too. [Henri Sivonen did it](https://hsivonen.fi/modern-cpp-in-rust/), but the way I read that article, it seems that the library structure was a bit helpful, but that the approach does not generalize.

> So my _hope_ would be that we can have a constructor that takes a `Locale`. Now, even that constructor imho should be fallible because we may not have data for that locale.

I think, instead we should have a constructor that can accept a `Locale`-like (i.e. a type with a trait bound).  We, of course, get to define what goes into that trait.

Reply:
Made the required methods fallible. PTAL.